larpandoracontent/LArThreeDReco/LArHitCreation/RANSACMethod.cc

```diff
@@ -0,0 +1,490 @@
+            hitsToAdd.size(), smallIterCount);
+
+        coherentHitCount += hitsToAdd.size();
+
+        // ATTN: If finished first pass, reset and run from opposite end.
+        //       If finished second (backwards) pass, stop.
+        if (!continueFitting && extendDirection == ExtendDirection::Forward)
+        {
+            extendDirection = ExtendDirection::Backward;
+            hitsToUseForFit.clear();
+            currentPoints3D.clear();
+            smallIterCount = 0;
+
+            auto it = sortedHits.begin();
+            // TODO: Randomly chosen "at least 5 fits worth", evaluate.
+            while (currentPoints3D.size() < (5 * 80) && it != sortedHits.end())
+            {
+                currentPoints3D.push_back(*it);
+                ++it;
+            }
+
+            std::reverse(currentPoints3D.begin(), currentPoints3D.end());
+            LArRANSACMethod::GetHitsForFit(currentPoints3D, hitsToUseForFit, 0, 0);
+        }
+        else if (!continueFitting && extendDirection == ExtendDirection::Backward)
+            break;
+    }
+
+    for (auto const& caloProtoPair : inlyingHitMap) {
+        finalHits.push_back(caloProtoPair.second);
+    }
+
+    // ATTN: This count is the count of every considered hit, not the hits that were added.
+    //       This allows distinguishing between models that fit to areas with more hits in.
+    return coherentHitCount;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool LArRANSACMethod::GetHitsForFit(
+        std::list<RANSACHit> &currentPoints3D,
+        RANSACHitVector &hitsToUseForFit,
+        const int addedHitCount,
+        int smallAdditionCount
+)
+{
+    const int HITS_TO_KEEP = 80; // TODO: Config and consolidate (reverse bit).
+    const int FINISHED_THRESHOLD = 10; // TODO: Config
+
+    // ATTN: Four options:
+    //  Added no hits at the end: Stop.
+    //  Added small number of hits at end repeatedly: Stop.
+    //  Added small number of hits inside fit: Clear and add hits.
+    //  Added lots of hits of hits: Don't add hits, just trim vector.
+    if (addedHitCount == 0 && currentPoints3D.size() == 0)
+        return false;
+
+    if (addedHitCount < 2 && smallAdditionCount > FINISHED_THRESHOLD)
+        return false;
+
+    if (addedHitCount <= 2 && currentPoints3D.size() != 0)
+    {
+        int i = 0;
+        auto it = currentPoints3D.begin();
+        while(i <= HITS_TO_KEEP && currentPoints3D.size() != 0)
+        {
+            hitsToUseForFit.push_back(*it);
+            it = currentPoints3D.erase(it);
+
+            if (hitsToUseForFit.size() >= HITS_TO_KEEP)
+                hitsToUseForFit.erase(hitsToUseForFit.begin());
+
+            ++i;
+        }
+    }
+    else if (addedHitCount > 0)
+    {
+        auto it = hitsToUseForFit.begin();
+        while (hitsToUseForFit.size() > HITS_TO_KEEP)
+            it = hitsToUseForFit.erase(it);
```

Comment:
Given you know how many elements need to be erased, I would think you could erase a set of hits in one call, rather than a loop?